domain/domain.go

```diff
@@ -1134,32 +1158,48 @@ func (do *Domain) checkReplicaRead(ctx context.Context, pdClient pd.Client) erro
 		}
 	}
 
-	enabled := false
-	// if stores don't have zone labels or are distribued in 1 zone, just disable cloeset replica read.
-	if len(storeZones) > 1 {
-		enabled = true
-		servers, err := infosync.GetAllServerInfo(ctx)
-		if err != nil {
-			return err
-		}
-		for _, s := range servers {
-			if v, ok := s.Labels[placement.DCLabelKey]; ok && v != "" {
-				if _, ok := storeZones[v]; !ok {
-					enabled = false
-					break
-				}
+	// no stores in this AZ
+	if _, ok := storeZones[zone]; !ok {
+		variable.SetEnableAdaptiveReplicaRead(false)
+		return nil
+	}
+
+	servers, err := infosync.GetAllServerInfo(ctx)
+	if err != nil {
+		return err
+	}
+	svrIdsInThisZone := make([]string, 0)
+	for _, s := range servers {
+		if v, ok := s.Labels[placement.DCLabelKey]; ok && v != "" {
+			if _, ok := storeZones[v]; ok {
 				storeZones[v] += 1
-			}
-		}
-		if enabled {
-			for _, count := range storeZones {
-				if count == 0 {
-					enabled = false
-					break
+				if v == zone {
+					svrIdsInThisZone = append(svrIdsInThisZone, s.ID)
 				}
 			}
 		}
 	}
+	enabledCount := math.MaxInt
+	for _, count := range storeZones {
+		if count < enabledCount {
+			enabledCount = count
+		}
+	}
+	// sort tidb in the same AZ by ID and disable the tidb with bigger ID
+	// because ID is unchangeable, so this is a simple and stable algorithm to select
+	// some instances across all tidb servers.
+	if enabledCount < len(svrIdsInThisZone) {
+		sort.Slice(svrIdsInThisZone, func(i, j int) bool {
+			return strings.Compare(svrIdsInThisZone[i], svrIdsInThisZone[j]) < 0
+		})
+	}
```

Comment:
Is there a possiblity of `enabledCount > len(svrIdsInThisZone)`? Then `svrIdsInThisZone[enabledCount:]` is invalid.